contracts/V4Router.sol

```diff
@@ -0,0 +1,189 @@
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+import {PathKey} from "./libraries/PathKey.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+
+/// @title UniswapV4Routing
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+abstract contract V4Router is IV4Router, ILockCallback {
+    using CurrencyLibrary for Currency;
+
+    IPoolManager immutable poolManager;
+
+    /// @dev Only the pool manager may call this function
+    modifier poolManagerOnly() {
+        if (msg.sender != address(poolManager)) revert NotPoolManager();
+        _;
+    }
+
+    constructor(IPoolManager _poolManager) {
+        poolManager = _poolManager;
+    }
+
+    function _v4Swap(SwapType swapType, bytes memory params) internal {
+        poolManager.lock(abi.encode(SwapInfo(swapType, msg.sender, params)));
+    }
+
+    /// @inheritdoc ILockCallback
+    function lockAcquired(bytes calldata encodedSwapInfo) external override poolManagerOnly returns (bytes memory) {
+        SwapInfo memory swapInfo = abi.decode(encodedSwapInfo, (SwapInfo));
+
+        if (swapInfo.swapType == SwapType.ExactInput) {
+            _swapExactInput(abi.decode(swapInfo.params, (IV4Router.ExactInputParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactInputSingle) {
+            _swapExactInputSingle(abi.decode(swapInfo.params, (IV4Router.ExactInputSingleParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactOutput) {
+            _swapExactOutput(abi.decode(swapInfo.params, (IV4Router.ExactOutputParams)), swapInfo.msgSender);
+        } else if (swapInfo.swapType == SwapType.ExactOutputSingle) {
+            _swapExactOutputSingle(abi.decode(swapInfo.params, (IV4Router.ExactOutputSingleParams)), swapInfo.msgSender);
+        } else {
+            revert InvalidSwapType();
+        }
+
+        return bytes("");
+    }
+
+    function _swapExactInputSingle(IV4Router.ExactInputSingleParams memory params, address msgSender) private {
+        _swapExactPrivate(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            msgSender,
+            true,
+            true,
+            params.hookData
+        );
+    }
+
+    function _swapExactInput(IV4Router.ExactInputParams memory params, address msgSender) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountOut;
+
+            for (uint256 i = 0; i < pathLength; i++) {
+                (PoolKey memory poolKey, bool zeroForOne) = _getPoolAndSwapDirection(params.path[i], params.currencyIn);
+                amountOut = uint128(
+                    -_swapExactPrivate(
+                        poolKey,
+                        zeroForOne,
+                        int256(int128(params.amountIn)),
```

Comment:
should just be able to cast to int256 directly

Reply:
think that's disallowed since [v0.8](https://docs.soliditylang.org/en/latest/080-breaking-changes.html):
>The conversion is only allowed when there is at most one change in sign, width or type-category (int, address, bytesNN, etc.). To perform multiple changes, use multiple conversions.